### PR TITLE
operators / ground operators

### PR DESCRIPTION
**Pull request by tomsilver**

### CODE REVIEW

File: src/structs.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def __hash__(self) -> int:
 
 
 @dataclass(frozen=True, order=True, repr=False)
-class TypedEntity:
+class _TypedEntity:
     """Struct defining an entity with some type, either an object (e.g.,
     block3) or a variable (e.g., ?block).
```
Comment (ronuchit) [inline src/structs.py:40]: Append "Should not be instantiated externally." to this docstring

File: src/structs.py [hunk 2]
```diff
@@ -222,10 +222,16 @@ def variables(self):
         return list(self.entities)
 
     @cached_property
-    def _str(self):
+    def _str(self) -> str:
         return (str(self.predicate) + "(" +
                 ", ".join(map(str, self.variables)) + ")")
 
+    def ground(self, sub) -> GroundAtom:
```
Comment (ronuchit) [inline src/structs.py:229]: Would we ever need to go the other way, i.e. GroundAtom to LiftedAtom? We would need this sometimes right, like when we lift effects for clustering?

Comment (ronuchit) [inline src/structs.py:229]: Please annotate the type of "sub" parameter

File: src/structs.py [hunk 3]
```diff
@@ -303,3 +309,92 @@ class _Option:
...
+    def __str__(self) -> str:
+        return self._str
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __hash__(self) -> int:
+        return self._hash
+
+    def __eq__(self, other) -> bool:
```
Comment (ronuchit) [inline src/structs.py:349]: Please annotate the type of "other" (actually, I think this is missing in a few places in this file)

File: src/structs.py [hunk 4]
```diff
@@ -303,3 +309,92 @@ class _Option:
...
+        delete_effects = {atom.ground(sub) for atom in self.delete_effects}
+        sampler = lambda s: self._sampler(s, objects)
+        return _GroundOperator(self, objects, preconditions, add_effects,
+                               delete_effects, self.option, sampler)
+
+
+@dataclass(frozen=True, repr=False, eq=False)
+class _GroundOperator:
+    """A ground operator is an operator + objects."""
+    operator: Operator
```
Comment (ronuchit) [inline src/structs.py:369]: Is there a reason GroundOperators need to store a pointer to the lifted Operator? I'm fine with it if we have a reason, but right now it looks like we're just using that lifted Operator's name, so we could just pass the name in instead?

File: tests/test_structs.py [hunk 5]
```diff
@@ -202,3 +203,54 @@ def _terminal(s, p):
     assert np.all(option.policy(state) == np.array(params)*2)
     assert not option.initiable(state)
     assert not option.terminal(state)
+
+
+def test_operators():
+    """Tests for Operators and _GroundOperators.
```
Comment (ronuchit) [inline tests/test_structs.py:209]: Operator and _GroundOperator [to avoid confusion]

File: src/structs.py [hunk 6]
```diff
@@ -303,3 +309,92 @@ class _Option:
...
+class Operator:
+    """Struct defining an operator (as in STRIPS). Lifted!
+    """
+    name: str
+    parameters: Sequence[Variable]
+    preconditions: Collection[LiftedAtom]
+    add_effects: Collection[LiftedAtom]
+    delete_effects: Collection[LiftedAtom]
+    option: ParameterizedOption
+    # A sampler maps a state and objects to a option parameters.
```
Comment (ronuchit) [inline src/structs.py:324]: typo: "to option parameters"

File: src/structs.py (hunk 2 above)
Comment (tomsilver) [reply src/structs.py:229]: so polite

Comment (tomsilver) [reply src/structs.py:229]: so polite

File: src/structs.py (hunk 3 above)
Comment (tomsilver) [reply src/structs.py:349]: this is apparently not recommended:

```
predicators/src/structs.py:356: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"
predicators/src/structs.py:356: note: This violates the Liskov substitution principle
predicators/src/structs.py:356: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
predicators/src/structs.py:356: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
predicators/src/structs.py:356: note:     def __eq__(self, other: object) -> bool:
predicators/src/structs.py:356: note:         if not isinstance(other, Operator):
predicators/src/structs.py:356: note:             return NotImplemented
predicators/src/structs.py:356: note:         return <logic to compare two Operator instances>
```

File: src/structs.py (hunk 4 above)
Comment (tomsilver) [reply src/structs.py:369]: I can't think of a reason but I would like to keep it

**Pull request merged**